fulltext/services/store.py

```diff
@@ -1,191 +1,189 @@
-                KeySchema=[
-                    {'AttributeName': 'document', 'KeyType': 'HASH'},
-                    {'AttributeName': 'created', 'KeyType': 'RANGE'}
-                ],
-                AttributeDefinitions=[
-                    {"AttributeName": 'document', "AttributeType": "S"},
-                    {"AttributeName": 'created', "AttributeType": "S"}
-                ],
-                ProvisionedThroughput={    # TODO: make this configurable.
-                    'ReadCapacityUnits': 20,
-                    'WriteCapacityUnits': 20
-                }
-            )
-            waiter = table.meta.client.get_waiter('table_exists')
-            waiter.wait(TableName=self.table_name)
-            logger.debug('Created table: %s' % self.table_name)
-        except ClientError as e:
-            logger.debug('Table already exists: %s' % self.table_name)
-
-    def create(self, document_id: str, content: str) -> None:
-        """
-        Store fulltext content for a document.
-
-        Parameters
-        ----------
-        sequence_id : int
-        state : str
-        document_id : str
-
-        Raises
-        ------
-        IOError
-        """
-        logger.debug('Store record for %s' % document_id)
-        entry = {
-            'document': document_id,
-            'version': self.version,
-            'created':  datetime.now().isoformat(),
-            'content': gzip.compress(bytes(content, encoding='utf-8'))
-        }
+
+class StorageFailed(RuntimeError):
+    """Could not store content."""
+
+
+class Storage(metaclass=MetaIntegration):
+    """Provides storage integration."""
+
+    def __init__(self, volume: str) -> None:
+        """Check and set the storage volume."""
+        if not os.path.exists(volume):
+            try:
+                os.makedirs(volume)
+            except Exception as e:
+                raise ConfigurationError("Cannot create storage volume") from e
+
+        self._volume = volume
+
+    def _paper_path(self, identifier: str, bucket: str) -> str:
+        return os.path.join(self._volume, bucket, identifier[:4], identifier)
+
+    def _path(self, identifier: str, version: str, content_fmt: str,
+              bucket: str) -> str:
+        return os.path.join(self._paper_path(identifier, bucket),
+                            version, content_fmt)
+
+    def _meta_path(self, identifier: str, version: str, bucket: str) -> str:
+        return os.path.join(self._paper_path(identifier, bucket),
+                            version, 'meta.json')
+
+    def _creation_time(self, path: str) -> datetime:
+        return datetime.fromtimestamp(os.path.getmtime(path), tz=UTC)
+
+    def _latest_version(self, identifier: str,
+                        bucket: str = SupportedBuckets.ARXIV) -> str:
+        def _try_float(value: str) -> float:
+            try:
+                return float(value)
+            except ValueError:
+                return 0.0
```

Comment:
Maybe put in a logger.debug("non-float version {value}")

Reply:
Added. Also put in this comment, because I'm not terribly happy with what's going on there:

```
# TODO: consider refactoring this to support full semantic versions, rather 
# than only major.minor that can be expressed as a float. This package 
# has decent support: https://pypi.org/project/semver/
```